### ISSUE TITLE
New function derivatives()

### ISSUE THREAD

**Issue by gavinsimpson** [gavinsimpson/gratia#13]
Add a new function `derivatives()` to compute finite-difference-based derivatives of fitted splines.

In effect this will soft-deprecate `fderiv()` as I would prefer a more tidy-friendly return object in keeping with the other functions in **gratia**.

* [x] `derivatives()` should also offer alternative ways of computing derivatives (forward, backward, or centred).
* [x] `derivatives()` should allow computing second derivatives (see #5 for an initial go at adding this functionality through `fderiv()`)
* [x] `derivatives()` should return a tibble with a variable indexing which smooth the derivatives are for
* [x] `derivatives()` should return confidence or simultaneous intervals
* [x] `draw()` method for derivatives

**Comment by grasshoppermouse** [comment]
What about a geom_dydx() to complement geom_smooth(). Thoughts?

**Comment by gavinsimpson** [comment]
Thanks @grasshoppermouse. I hadn't really thought about *gratia* including geoms for ggplot - I suspect a `stat` is more in keeping with the design idea, unless you were thinking of some specific graphical display for derivatives.

I;m not clear on what such a geom would display? As I'm really only concerning myself with GAMs in *gratia* the package isn't (or at least at the moment) isn't intended to work with other types of models that `geom_smooth()` is typically used for. But if you have further detail to add as to what this geom would do/show I will certainly consider it.

**Comment by grasshoppermouse** [comment]
Well, I haven't thought it out too carefully, but geom_smooth estimates and displays a function y(x) from data. I think it defaults to GAM for >1000 data points.

A geom_dydx would first estimate y(x) using a GAM, and then display y’(x). If find myself doing this a lot by hand. It obviously doesn’t play well with Hadley’s aversion to different scales for the y-axis, however, if you want to display the original data points too. 

I guess a stat is a better suggestion.

**Comment by gavinsimpson** [comment]
I wondered if that was where your thinking was at, and it's where I'm getting a little hung up at the moment. Even the stat idea isn't that intuitive as you'd only want the user to supply `aes(x = xvar)` and then work out what the yvar should be.

I'd like to make it easier to plot derivatives -- the `draw()` method when I get to it should make it easy to plot the derivatives themselves.

In quite a few of my papers or in progress work where I've used derivatives I've actually plotted the estimated smooth and then superimposed on that smooth some information about the derivative (e.g. where 95% CI on derivative excludes 0) or used a diverging palette and set the colour aesthetic of the line to the derivative variable. For second derivatives I've only really seen the turning points indicated by dots on the fitted smoother.
